### PR TITLE
VOS loss

### PR DESCRIPTION
**Pull request by tla93**
add VOS loss, detector and segmentation example
pre-commit.config : update flake8 version to 5.0.2

### CODE REVIEW

File: src/pytorch_ood/detector/vos.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    def fit_features(self: Self, *args, **kwargs) -> Self:
+        """
+        Not required.
+        """
+        return self
+
+    def __init__(self, model: torch.nn.Module, weights_energy: torch.nn.Module):
+        """
+        :param t: Temperature value :math:`T`. Default is 1.
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/vos.py]: documentation inconsistent

File: src/pytorch_ood/detector/vos.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return self.score(logits)
+
+    @staticmethod
+    def score(logits: torch.Tensor, weights_energy: torch.nn.Module) -> torch.Tensor:
+        """
+        :param logits: logits of input
+        :param weights_energy: energy weights as torch.nn.module
+        """
+        # Permutation depends on shape of logits
+        tmp_scores_ = logits.permute(0, 2, 3, 1)
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/vos.py]: will this only work in segmentation settings?

File: src/pytorch_ood/loss/vos.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    :see Implementation:
+        `GitHub <https://github.com/deeplearning-wisc/vos/>`__
+
+    For initialisation of :math:`\\phi` and :math:`weights\\_energy`  see `GitHub <https://github.com/deeplearning-wisc/vos/blob/a449b03c7d6e120087007f506d949569c845b2ec/classification/CIFAR/train_virtual.py#L132>`__ .
+
+    Notice, that this loss is more effective with scheduler and low (e.g 0.001) learningrate see `GitHub <https://github.com/deeplearning-wisc/vos/blob/a449b03c7d6e120087007f506d949569c845b2ec/classification/CIFAR/train_virtual.py#L152>`__.
+    """
+
+    def __init__(
+        self, logistic_regression, weights_energy, alpha=0.1, device="cuda:0", reduction="mean"
```
Comment (kkirchheim) [inline src/pytorch_ood/loss/vos.py]: please add type hints to parameters

File: examples/segmentation/street_vos.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    encoder_weights="imagenet",
+    in_channels=3,
+    classes=num_classes,
+).to(device)
+
+# %%
+# Create non-linear neural network functions
+phi = torch.nn.Linear(1, 2)
+weights_energy = torch.nn.Linear(num_classes, 1)
+if 'cuda' in device:
```
Comment (kkirchheim) [inline examples/segmentation/street_vos.py]: please use to .to(device) method here as well

**Review by kkirchheim — CHANGES_REQUESTED** [review]

File: examples/segmentation/street_vos.py (hunk 4 above)
Comment (kkirchheim) [reply examples/segmentation/street_vos.py]: also, these functions are linear

File: .gitignore [hunk 5]
```diff
@@ -254,3 +254,13 @@ dmypy.json
 cython_debug/
 
 # End of https://www.toptal.com/developers/gitignore/api/python,intellij
+.github/workflows/python-package.yml
+.vscode/settings.json
+.pre-commit-config.yaml
+.pre-commit-config.yaml
+.gitignore
```
Comment (kkirchheim) [inline .gitignore:258]: why are these files on the gitignore list?

File: src/pytorch_ood/loss/vos.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            logits_form = logits
+
+        # code snippet from https://github.com/deeplearning-wisc/vos/blob/a449b03c7d6e120087007f506d949569c845b2ec/classification/CIFAR/train_virtual.py#L245
+        energy_x_in = self._energy(logits_form[is_known(y)])
+        energy_v_out = self._energy(logits_form[is_unknown(y)])
+
+        input_for_lr = torch.cat((energy_x_in, energy_v_out), -1)
+        if "cuda" in self.device:
+            labels_for_lr = torch.cat(
+                (torch.ones(len(energy_x_in)).cuda(), torch.zeros(len(energy_v_out)).cuda()), -1
```
Comment (kkirchheim) [inline src/pytorch_ood/loss/vos.py]: please use .to(device) function

File: src/pytorch_ood/loss/vos.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            )
+        else:
+            labels_for_lr = torch.cat(
+                (torch.ones(len(energy_x_in)), torch.zeros(len(energy_v_out))), -1
+            )
+
+        criterion = torch.nn.CrossEntropyLoss(reduction=self.reduction)
```
Comment (kkirchheim) [inline src/pytorch_ood/loss/vos.py]: please construct cross-entropy in constructor and not on each update

File: tests/detectors/test_vos.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import unittest
+
+import torch
+
+from src.pytorch_ood.detector import VOSBased
```
Comment (kkirchheim) [inline tests/detectors/test_vos.py]: this should fail since detector was renamed

File: tests/loss/test_vos.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+from src.pytorch_ood.loss import VOSRegLoss
+from tests.helpers.model import SegmentationModel
+
+torch.manual_seed(123)
+
+
+class TestEnergyRegularization(unittest.TestCase):
+    """
+    Test code for energy bounded learning
```
Comment (kkirchheim) [inline tests/loss/test_vos.py]: not up to date

File: src/pytorch_ood/detector/webo.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        if self.model is None:
+            raise ModelNotSetException
+
+        return self.score(self.model(x), self.weights)
+
+    def predict_features(self, logits: Tensor) -> Tensor:
+        """
+        :param logits: logits given by your model
+        """
+        return self.score(logits)
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/webo.py]: this will fail. please write some more tests.
